File: nodejs/src/state/OrderState.ts
// Interface tối giản để tránh circular dependency với Order
export interface IOrder {
    id: string;
    setState(state: OrderState): void;
}

abstract class OrderState {
    protected order: IOrder;

    constructor(order: IOrder) {
        this.order = order;
    }

    abstract getName(): string;
    abstract next(): void;
    abstract cancel(): void;

    addItem(item: { getDescription(): string }): boolean {
        console.log(`❌ Không thể thêm đồ ở trạng thái ${this.getName()}`);
        return false;
    }
}

export default OrderState;

File: spring/demo/src/main/java/com/example/demo/state/OrderState.java
package com.example.demo.state;

import com.example.demo.decorator.Beverage;
import com.example.demo.model.Order;

public abstract class OrderState {
    protected final Order order;

    protected OrderState(Order order) {
        this.order = order;
    }

    public abstract String getName();

    public abstract void next();

    public abstract void cancel();

    public boolean addItem(Beverage item) {
        System.out.println("Không thể thêm đồ ở trạng thái " + getName());
        return false;
    }
}
